### PR TITLE
[WIP] Implement Mbed TLS 3 semantics of mbedtls_ssl_write()

### PR DESCRIPTION
**Pull request by hanno-arm**
This is a work-in-progress PR aiming to implement the proposed new semantics of `mbedtls_ssl_write()` for Mbed TLS 3 described in #4183.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
There are merge conflicts that need to be resolved before a code review.

Are you waiting for a design review? Does this need to be in Mbed TLS 3.0?

**Comment by hanno-arm** [comment]
@mpg @gilles-peskine-arm Any chance we can still discuss this for 3.0? @gilles-peskine-arm I believe you have already conducted a first design study as part of https://github.com/ARMmbed/mbedtls/issues/4184.

**Comment by mpg** [comment]
This is still on my list but so far there are items that are higher on my list. It's hard right now to estimate if/when I'll get to this, sorry.

**Comment by hanno-arm** [comment]
@mpg That's OK. In the meantime I will ensure that everything passes.

**Comment by gilles-peskine-arm** [comment]
My priority order is (1) security relevant stuff, (2) re-review or second review for 3.0, (3) other 3.0 item. At the moment I still have a bit of room for some (3) next week, so this PR is on my queue, but I can't promise that I'll reach it.

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
A few small things, looks good otherwise.

File: library/ssl_msg.c [hunk 1]
```diff
@@ -5505,6 +5519,37 @@ int mbedtls_ssl_write( mbedtls_ssl_context *ssl, const unsigned char *buf, size_
...
+    }
+    max_len = (size_t) ret;
+
+    if( len > max_len )
+    {
+#if defined(MBEDTLS_SSL_PROTO_DTLS)
+        if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 1, ( "fragment larger than the (negotiated) "
+                                        "maximum fragment length: %zu > %zu",
```
Comment (davidhorstmann-arm) [inline library/ssl_msg.c:5544]: To fix the CI issues, caused by `"%z"` being unsupported on some platforms:
```suggestion
                                        "maximum fragment length: %" MBEDTLS_PRINTF_SIZET
                                        " > %" MBEDTLS_PRINTF_SIZET,
```

File: include/mbedtls/ssl.h [hunk 2]
```diff
@@ -3851,25 +3851,35 @@ int mbedtls_ssl_renegotiate( mbedtls_ssl_context *ssl );
...
- * \param ssl      SSL context
- * \param buf      buffer holding the data
- * \param len      how many bytes must be written
+ * \warning        This function does not send any of the application data
+ *                 provided by the user, but only prepares and queues the
+ *                 corresponding protected application data records. Dispatch to
+ *                 the underlying transport is only attempted when the user
+ *                 explicitly calls mbedtls_ssl_flush(), or when a subsequent
+ *                 call to mbedtls_ssl_write() can not make progress without
+ *                 dispatching some queued application records first.
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h:3866]: This seems contradictory with “This function does not send any of the application data” above. Can `mbedtls_ssl_write` cause application data to be written or not?

(Generally https://github.com/ARMmbed/mbedtls/issues/4183 is a lot clearer than the documentation here.)

File: include/mbedtls/ssl.h [hunk 3]
```diff
@@ -3851,25 +3851,35 @@ int mbedtls_ssl_renegotiate( mbedtls_ssl_context *ssl );
...
  *
- * \return         The (non-negative) number of bytes actually written if
- *                 successful (may be less than \p len).
+ * \param ssl      The SSL context to use.
+ * \param buf      The buffer holding the applicatoin data to be queued.
+ * \param len      The size of \p buf in bytes.
+ *
+ * \return         The (non-negative) number of bytes queued for dispatch
+ *                 if successful (may be less than \p len). In this case,
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h:3873]: When would this function return a value that is less than `len`?

File: include/mbedtls/ssl.h [hunk 4]
```diff
@@ -3908,10 +3927,44 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
...
+/**
+ * \brief          Dispatch all previously queued application data
+ *
+ * \param ssl      The SSL context to use.
+ *
+ * \return         \c 0 if all data previously queued for dispatch via
+ *                 mbedtls_ssl_write() has been passed to the underlying
+ *                 transport.
+ * \return         #MBEDTLS_ERR_SSL_WANT_READ or #MBEDTLS_ERR_SSL_WANT_WRITE
+ *                 if the underlying transport was not available.
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h:3943]: This would benefit from an explanation of how `WANT_READ` can result from an attempt to write (repeat of the explanation given for `mbedtls_ssl_write`).

Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h:3943]: Please explain what `WANT_WRITE` indicates: it's possible that some data was written. Relate this to the blocking behavior (which controls when this function stops writing).

File: include/mbedtls/ssl.h [hunk 5]
```diff
@@ -3851,25 +3851,35 @@ int mbedtls_ssl_renegotiate( mbedtls_ssl_context *ssl );
...
  *
- * \warning        This function will do partial writes in some cases. If the
- *                 return value is non-negative but less than length, the
- *                 function must be called again with updated arguments:
- *                 buf + ret, len - ret (if ret is the return value) until
- *                 it returns a value equal to the last 'len' argument.
+ * \warning        This function may queue less data for dispatch than requested.
+ *                 If the return value is non-negative but less than \p len,
+ *                 the function should be called again for the remaining data.
  *
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h:3859]: Please explain the blocking behavior of this function. In particular, is it blocking when the TLS socket is in blocking mode?

File: include/mbedtls/ssl.h [hunk 6]
```diff
@@ -3908,10 +3927,44 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
...
  * \note           Attempting to write 0 bytes will result in an empty TLS
- *                 application record being sent.
+ *                 application record being prepared.
  */
 int mbedtls_ssl_write( mbedtls_ssl_context *ssl, const unsigned char *buf, size_t len );
 
+/**
+ * \brief          Dispatch all previously queued application data
+ *
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h:3936]: Please describe the blocking behavior of this function.

File: include/mbedtls/ssl.h [hunk 7]
```diff
@@ -3908,10 +3927,44 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
  *                 query the active maximum fragment length.
  *
  * \note           Attempting to write 0 bytes will result in an empty TLS
- *                 application record being sent.
+ *                 application record being prepared.
  */
 int mbedtls_ssl_write( mbedtls_ssl_context *ssl, const unsigned char *buf, size_t len );
 
+/**
+ * \brief          Dispatch all previously queued application data
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h:3935]: Not _all_ data, since `WANT_WRITE` is possible. It just sends what it can. To send all data, you need to call `flush` in a loop until it stops returning a “try again” error.

File: include/mbedtls/ssl.h [hunk 8]
```diff
@@ -3908,10 +3927,44 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
...
+ *                 #MBEDTLS_ERR_SSL_WANT_WRITE,
+ *                 #MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS or
+ *                 #MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS,
+ *                 you must stop using the SSL context for reading or writing,
+ *                 and either free it or call \c mbedtls_ssl_session_reset()
+ *                 on it before re-using it for a new connection; the current
+ *                 connection must be closed.
+ */
+int mbedtls_ssl_flush( mbedtls_ssl_context *ssl );
+
```
Comment (gilles-peskine-arm) [inline include/mbedtls/ssl.h:3967]: `while ret is WANT_xxx or xxx_IN_PROGRESS: flush` is such a common idiom that it should be a library function. Maybe call that `flush`, and rename the flush-what-you-can do `send`? Or `flush` and `flush_all`? Or `flush` with an `all` boolean flag?

File: programs/ssl/ssl_client2.c [hunk 9]
```diff
@@ -2417,6 +2417,19 @@ int main( int argc, char *argv[] )
     mbedtls_printf( " %d bytes written in %d fragments\n\n%s\n",
                     written, frags, (char *) buf );
 
+    /* Make sure all data is flushed. */
+    while( 1 )
+    {
+        ret = mbedtls_ssl_flush( &ssl );
+        if( ret != MBEDTLS_ERR_SSL_WANT_WRITE )
```
Comment (gilles-peskine-arm) [inline programs/ssl/ssl_client2.c:2424]: This program supports renegotiation so other errors are possible. Likewise for `ssl_server2`.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
The documentation needs work. Please clarify the documentation of the functions — it's especially not clear when `write` will flush data. Also, a changelog entry and a migration guide entry are necessary.

I think that the implementation is correct, based on the semantics described in https://github.com/ARMmbed/mbedtls/issues/4183, but I will need to do another pass once the edge cases are clarified.

I want to have unit tests that distinguish the old behavior from the new behavior, but this can be delayed until after the release.

File: include/mbedtls/ssl.h (hunk 2 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h:3866]: Good point. `mbedtls_ssl_write()` may try to deliver application data from previous calls to `mbedtls_ssl_write()`, but it will _never_ attempt to deliver application data from the current call. This way, you always know what a WANT_READ/WRITE referred to previous calls, not the current one, and that the current one therefore has to be retried.

Comment (hanno-arm) [reply include/mbedtls/ssl.h:3866]: I will clarify the documentation.

File: include/mbedtls/ssl.h (hunk 7 above)
Comment (hanno-arm) [reply include/mbedtls/ssl.h:3935]: That's true, but for the sake of the brief description, don't we usally omit the "Try to ..." prefix and explain the ways a function can fail in the main body of the documentation?

File: programs/ssl/ssl_client2.c (hunk 9 above)
Comment (hanno-arm) [reply programs/ssl/ssl_client2.c:2424]: We don't check for this in the previous version of the code, so I suggest we leave it for 3.X to improve the example programs in this regard.

File: include/mbedtls/ssl.h (hunk 7 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/ssl.h:3935]: For a function that either does the job or fails in a non-recoverable way, yes. But a `WANT_WRITE` return here is really a partial success: some progress has been made and normal operation can continue. So it's misleading to only refer to the complete success case.

File: programs/ssl/ssl_client2.c (hunk 9 above)
Comment (gilles-peskine-arm) [reply programs/ssl/ssl_client2.c:2424]: We do: the loop around `mbedtls_ssl_write` allows `MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS` and `MBEDTLS_ERR_SSL_WANT_READ` as well as `MBEDTLS_ERR_SSL_WANT_WRITE`.

**Comment by mpg** [comment]
It looks like there are already two active reviewers - @gilles-peskine-arm and @davidhorstmann-arm - so I'm removing myself as a reviewer, as well as removing the "needs: reviewer" label.

**Comment by mniestroj** [comment]
Is there any plan to continue work on that PR in the nearest future? Is there a chance to introduce behavior from this PR using new API (e.g. `mbedtls_ssl_write_noflush()`) for mbedTLS v3, so that API is not broken and "gather" write functionality is still introduced?

**Comment by tom-daubney-arm** [comment]
We are now converting older PRs to draft PRs where the following conditions are met: They have not been updated in the last 3 months, and they need more than non-trivial work to complete.
